pkg/containertools/untarer.go

```diff
@@ -0,0 +1,88 @@
+package containertools
+
+import (
+	"archive/tar"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/sirupsen/logrus"
+)
+
+// untarer can untar tar files.
+type untarer struct {
+	log *logrus.Entry
+}
+
+func newUntarer(logger *logrus.Entry) *untarer {
+	return &untarer{
+		log: logger,
+	}
+}
+
+func (u *untarer) Untar(ctx context.Context, reader *tar.Reader, path string) error {
+	var err error
+	path, err = filepath.Abs(path)
+	if err != nil {
+		return err
+	}
+
+	for {
+		header, err := reader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		// Determine proper file path info
+		info := header.FileInfo()
+		name := header.Name
+		abs := filepath.Join(path, name)
+
+		// If a dir, create it, then go to next segment
+		if info.Mode().IsDir() {
+			if err := os.MkdirAll(abs, os.ModePerm); err != nil {
+				u.log.Debugf("creating %s dir", abs)
+				return err
+			}
+			continue
+		}
+
+		// Create new file with custom file permissions
+		file, err := os.OpenFile(
+			abs,
+			os.O_RDWR|os.O_CREATE|os.O_TRUNC,
+			os.ModePerm,
+		)
+		if err != nil {
+			return err
+		}
+
+		u.log.Debugf("writing %s to disk", abs)
+		n, err := io.Copy(file, reader)
+		if err != nil {
+			return err
+		}
+
+		if err := file.Close(); err != nil {
```

Comment:
Does this need to live as a defer statement instead closer to the call site? It looks like we could leak file descriptors when the io.Copy call returns an error.

Reply:
Fair point.

I moved this whole section into a helper method so we can ensure the close happens before we copy the next file (we don't want more than one open at a time).